design/local-node-agent.md

```diff
@@ -0,0 +1,146 @@
+Remember that the attach operation will be initiated on the desired node's Rook agent by the creation of a volume attach CRD.
+When the agent performs the attach on its local node, it can update the CRD with information stating that the volume is now locked.
+Thus, the CRD itself is providing the accounting for the volume's lock in a generalized way, independent of the of the underlying storage protocol the agent chose to use for the volume.
+Of course, this lock can be augmented at the lower layer of the specific storage protocol, e.g., `rbd lock add` for RBD.
+
+Similarly, during the detach operation, the CRD will be deleted to signify that the volume has been unlocked.
+Since the Rook Flexvolume driver implements the attach/detach interface, detach will be called on the Controller manager node.
+If the node where the volume has attached has died, then this central driver/agent will need to perform the unlocking itself.
+This can be done by executing the storage protocol specific command (e.g., `rbd lock remove`) and the deletion of the volume attach CRD.
+
+#### Security
+There are two interfaces for communicating with and invoking operations on the Rook agent:
+1. Unix domain socket
+1. CRD events
+
+The Unix domain socket will have read/write only accessible by `root` and it is only accessible on the local node (not remotely accessible).
+
+CRD events (such as the creation of the volume attach CRD) will be using the standard Kubernetes security enforcement with RBAC.  Only cluster roles that are explicitly granted permissions to `GET`, `CREATE`, `DELETE`, etc. will be allowed to do so.
+
+### **Cluster Cleanup**
+Rook makes storage as a service a deeply integrated part of the Kubernetes cluster as opposed to an external entity.
+This integration makes more attention to lifecycle management of the storage components necessary.
+
+#### Hung RBD Kernel Module
+For example, if the monitor pods of a Rook cluster are no longer accessible while block storage is mapped to a node, the kernel RBD module will be hung and require a [power cycle of the machine](https://github.com/rook/rook/issues/376#issuecomment-318803799).
+
+The Rook agent can help mitigate this scenario, by watching for the cluster CRD delete event.
+When a Rook cluster is being deleted, there may still be consumers of the storage in the cluster in the form of pods with PVCs.
+When a Rook agent receives a cluster CRD delete event, they will respond by checking for any Rook storage on the local node they are running on and then forcefully remove them.
+
+To forcefully remove the storage from local pods, the agent will perform the following sequence of steps for each Rook PVC:
+```
+$ kubectl delete pvc <pvc name>
+$ sudo rbd unmap -o force /dev/rbdX
+# wait for it to time out or send SIGINT
+$ kubectl delete pv <pv name>
+$ sudo umount <mount point>
+```
+
+#### HostPath Data
```

Comment:
Local storage essentially brings storage on a local dir/device under kubernetes management. This means that during cleanup of a cluster we can delete the storage as we are cleaning up all the cluster resources. So when you create the next cluster there wouldn't be any cruft left on disk to cleanup, like you would with hostPath usage.